### DISCUSSION TITLE
kotlin-validation-pitfalls/

### DISCUSSION THREAD

**Discussion by giscus** [Hyeon9mak/Hyeon9mak.github.io#171] [Blog comments]
# kotlin-validation-pitfalls/

Kotlin 사용 환경에서 @NotBlank, @Size 등과 같은 javax.Validation annotation 을 사용하는 경우 제대로 validation 이 동작하지 않는 경우가 있다. 예시를 통해 문제 상황을 인지하고, 문제 원인과 해결 방법을 알아보자.  

https://hyeon9mak.github.io/kotlin-validation-pitfalls/

**Comment by wannte** [comment]
매번 글 잘 읽고 있습니다 ! 한 가지 질문 남깁니다.
Int, Long 같은 경우에는 java의 int, long으로 변환되서 0으로 setting되는 부분은 어떻게 처리하는 지 궁금합니다 ㅎㅎ

Int?, Long? 으로 받고 field:Notnull && 사용하는 시점에 !!  처리를 할까요?

**Reply by Hyeon9mak** [reply]
완태~ 매번 감사해요 :smile: :+1:
말씀해주신 부분 재현해보고 싶어서 여러가지 테스트를 해보았는데, 예상과 다르게 동작하더라구요.
Int 와 Long 을 non-null 하게 가진 data class 를 만들면 int, long 으로 치환되고
반대로 nullable 하게 가진 data class 를 만들면 Integer, Long 으로 치환되네요.

```kotlin
data class Book(
    val integerValue: Int,
    val longValue: Long,
)

public final class Book {
   private final int integerValue;
   private final long longValue;

   public Book(int integerValue, long longValue) {
      this.integerValue = integerValue;
      this.longValue = longValue;
   }
}
```

```kotlin
data class Book(
    val integerValue: Int? = null,
    val longValue: Long? = null,
)

public final class Book {
   @Nullable
   private final Integer integerValue;
   @Nullable
   private final Long longValue;

   public Book(@Nullable Integer integerValue, @Nullable Long longValue) {
      this.integerValue = integerValue;
      this.longValue = longValue;
   }
}
```

때문에 자동으로 0으로 치환되는 테스트를 못해보았어요.
혹시 재현 가능한 예시가 있을까요?

**Reply by wannte** [reply]
```kotlin
data class Book(
    val integerValue: Int,
    val longValue: Long,
)
```

아 의도한 상황은, 위의 상황에서 클라이언트가 아래의 요청을 보냈을 때,  `longValue` 를 보내지 않았음에도 불구하고, 0으로 받게되는 부분을 의미했어요! 
```json
{
  integerValue: 1
}
```

---
```kotlin
data class Book(
    @feild:NotNull
    val integerValue: Int,
    @feild:NotNull
    val longValue: Long,
)
```
와 같은 방식으로 처리했을 때, 클라이언트가 integerValue, longValue 를 요청에 포함하는 validation이 적용되기를 기대했으나, `{}` 요청을 보내더라도 전부 0으로 세팅되는 경험을 했어서 남겼던 코멘트 였습니다 :)

**Comment by be-student** [comment]
코틀린에서 저런 부분이 있군요
데이터 클래스에서 예상치 못한 방식의 작동이 있다면 생성자를 생각해봐라
감사합니다
